### PR TITLE
[A1] Enhancement: `ls`

### PR DESCRIPTION
**Pull request by Rocketeer8**
# POSIX Feature Implementation for `ls`
  
  ## Introduction
  This document details the implementation of additional POSIX-compliant features in the `ls` command for xv6-riscv. The new features include:
  - `-a`: Show hidden files
  - `-F`: Append file type symbols
  - `-l`: Display long format listing
  - `-m`: Output entries in a comma-separated format
  
  Each section below outlines the implementation, accompanied by relevant code snippets and test cases.
  
  ---
  
  ## Implemented Features and Code Changes
  
  ### 1. `-a` (Show Hidden Files)
  #### **Implementation**
  By default, files starting with `.` are hidden. The flag `-a` allows these files to be listed.
  
  **Modified `ls` function:**
  ```c
  if (!show_hidden && de.name[0] == '.') {
      continue; // Skip hidden files unless -a is enabled
  }
  ```
  **Flag Parsing in `main()`:**
  ```c
  else if (argv[i][j] == 'a') show_hidden = 1; // Enable -a flag
  ```
  **Test Cases:**
  -  `ls` → Should exclude hidden files
  -  `ls -a` → Should include hidden files like `.bashrc`
  -  `ls -a /some/directory` → Should list hidden and normal files
  
  ---
  
  ### 2. `-F` (Append File Type Symbols)
  #### **Implementation**
  Appends file type indicators: `/` for directories, `|` for devices.
  
  **Function to append symbols:**
  ```c
  void append_file_symbol(char *name, short type) {
      switch (type) {
          case T_DIR:
              printf("%s/\n", name);
              break;
          case T_FILE:
              printf("%s\n", name);
              break;
          case T_DEVICE:
              printf("%s|\n", name);
              break;
          default:
              printf("%s\n", name);
      }
  }
  ```
  **Flag Parsing in `main()`:**
  ```c
  else if (argv[i][j] == 'F') append_symbols = 1; // Enable -F flag
  ```
  **Test Cases:**
  -  `ls -F` → Should display directories with `/`, devices with `|`
  -  `ls -a -F` → Should show hidden files with appropriate type symbols
  
  ---
  
  ### 3. `-l` (Long Listing Format)
  #### **Implementation**
  Displays detailed file metadata in a format similar to POSIX. This is a simplified version without the "Group", "Owner", and "Date and Time" fields. 
  
  **Modified `ls` function:**
  ```c
  if (long_format) {
      char mode[11] = "----------"; // Dummy file mode
      mode[0] = (st.type == T_DIR) ? 'd' : '-';
      printf("%s %u ? ? %u ??? %s\n", mode, 1, (uint)st.size, fmtname(buf));
  }
  ```
  **Flag Parsing in `main()`:**
  ```c
  else if (argv[i][j] == 'l' && stream_output == 0) long_format = 1; // Enable -l flag
  ```
  **Test Cases:**
  -  `ls -l` → Should display file permissions, size, and name
  -  `ls -l -a` → Should list hidden files with metadata
  -  `ls -l -F` → Should show file details along with type indicators
  
  ---
  
  ### 4. `-m` (Comma-Separated Output)
  #### **Implementation**
  Prints entries in a single line, separated by commas.
  
  **Modified `ls` function:**
  ```c
  if (stream_output) { // -m flag: Format as comma-separated list
      if (!first_entry) {
          printf(", "); // Add comma before entries except the first one
      }
      printf("%s", fmtname(buf));
      first_entry = 0; // Mark that at least one entry has been printed
  }
  ```
  **Flag Parsing in `main()`:**
  ```c
  else if (argv[i][j] == 'm') {
      stream_output = 1;
      long_format = 0; // -m flag disables -l flag
  }
  ```
  **Test Cases:**
  -  `ls -m` → Should print files in a single line, comma-separated
  -  `ls -m -a` → Should include hidden files in the comma-separated list
  
  ---
  
  ## Edge Cases and Misuse Handling
  ### **Invalid Combinations**
  - `ls -m -l`: The `-m` flag disables `-l`, and the implementation ensures that the output format is adjusted accordingly.
  - `ls -F -m`: Ensures that type indicators are appended while maintaining comma separation.
  
  ### **Long Paths and Memory Constraints**
  - **Handled Path Length:** Ensures paths do not exceed buffer limits to prevent crashes.
  - **Memory Efficiency:** Limits directory entries processed to `MAX_ENTRIES = 100`.

### CODE REVIEW

File: user/ls.c [hunk 1]
```diff
@@ -4,84 +4,173 @@
...
+  // **Parse flags**
+  for (; i < argc; i++) {
+    if (argv[i][0] == '-') {
+      for (int j = 1; argv[i][j] != '\0'; j++) {
+        if (argv[i][j] == 'a') show_hidden = 1; // **Enable -a flag**
+        else if (argv[i][j] == 'F') append_symbols = 1; // **Enable -F flag**
+        else if (argv[i][j] == 'l') long_format = 1; // **Enable -l flag**
+        else if (argv[i][j] == 'm') { // **Enable -m flag**
+          stream_output = 1;
+          long_format = 0; // **-m flag disables -l flag**
```
Comment (Nish7) [inline user/ls.c:169]: ordering of the -l after the -m could cause conflicts. as we are disabling long format, might wanna add a condition within the long format (line 155) to check for stream output. does that make sense?

File: user/ls.c [hunk 2]
```diff
@@ -4,84 +4,173 @@
...
+        }
+      }
+    } else {
+      break;
```
Comment (Nish7) [inline user/ls.c]: rather than breaking. i believe we should exit with an appropriate message as well.

File: user/ls.c (hunk 1 above)
Comment (Rocketeer8) [reply user/ls.c:169]: Thanks for the comment. I added `argv[i][j] == 'l' && stream_output == 0` for more robust code. 

Though ls -ml works the same as ls -m already by ignoring the -l flag so i think it's fine.

File: user/ls.c (hunk 2 above)
Comment (Rocketeer8) [reply user/ls.c]: The break is for ending the flag settings and moving onto the ls operations, I will leave a comment

**Comment by Rocketeer8** [comment]
# POSIX Feature Implementation for `ls`

## Introduction
This document details the implementation of additional POSIX-compliant features in the `ls` command for xv6-riscv. The new features include:
- `-a`: Show hidden files
- `-F`: Append file type symbols
- `-l`: Display long format listing
- `-m`: Output entries in a comma-separated format

Each section below outlines the implementation, accompanied by relevant code snippets and test cases.

---

## Implemented Features and Code Changes

### 1. `-a` (Show Hidden Files)
#### **Implementation**
By default, files starting with `.` are hidden. The flag `-a` allows these files to be listed.

**Modified `ls` function:**
```c
if (!show_hidden && de.name[0] == '.') {
    continue; // Skip hidden files unless -a is enabled
}
```
**Flag Parsing in `main()`:**
```c
else if (argv[i][j] == 'a') show_hidden = 1; // Enable -a flag
```
**Test Cases:**
-  `ls` → Should exclude hidden files
-  `ls -a` → Should include hidden files like `.bashrc`
-  `ls -a /some/directory` → Should list hidden and normal files

---

### 2. `-F` (Append File Type Symbols)
#### **Implementation**
Appends file type indicators: `/` for directories, `|` for devices.

**Function to append symbols:**
```c
void append_file_symbol(char *name, short type) {
    switch (type) {
        case T_DIR:
            printf("%s/\n", name);
            break;
        case T_FILE:
            printf("%s\n", name);
            break;
        case T_DEVICE:
            printf("%s|\n", name);
            break;
        default:
            printf("%s\n", name);
    }
}
```
**Flag Parsing in `main()`:**
```c
else if (argv[i][j] == 'F') append_symbols = 1; // Enable -F flag
```
**Test Cases:**
-  `ls -F` → Should display directories with `/`, devices with `|`
-  `ls -a -F` → Should show hidden files with appropriate type symbols

---

### 3. `-l` (Long Listing Format)
#### **Implementation**
Displays detailed file metadata in a format similar to POSIX.

**Modified `ls` function:**
```c
if (long_format) {
    char mode[11] = "----------"; // Dummy file mode
    mode[0] = (st.type == T_DIR) ? 'd' : '-';
    printf("%s %u ? ? %u ??? %s\n", mode, 1, (uint)st.size, fmtname(buf));
}
```
**Flag Parsing in `main()`:**
```c
else if (argv[i][j] == 'l' && stream_output == 0) long_format = 1; // Enable -l flag
```
**Test Cases:**
-  `ls -l` → Should display file permissions, size, and name
-  `ls -l -a` → Should list hidden files with metadata
-  `ls -l -F` → Should show file details along with type indicators

---

### 4. `-m` (Comma-Separated Output)
#### **Implementation**
Prints entries in a single line, separated by commas.

**Modified `ls` function:**
```c
if (stream_output) { // -m flag: Format as comma-separated list
    if (!first_entry) {
        printf(", "); // Add comma before entries except the first one
    }
    printf("%s", fmtname(buf));
    first_entry = 0; // Mark that at least one entry has been printed
}
```
**Flag Parsing in `main()`:**
```c
else if (argv[i][j] == 'm') {
    stream_output = 1;
    long_format = 0; // -m flag disables -l flag
}
```
**Test Cases:**
-  `ls -m` → Should print files in a single line, comma-separated
-  `ls -m -a` → Should include hidden files in the comma-separated list

---

## Edge Cases and Misuse Handling
### **Invalid Combinations**
- `ls -m -l`: The `-m` flag disables `-l`, and the implementation ensures that the output format is adjusted accordingly.
- `ls -F -m`: Ensures that type indicators are appended while maintaining comma separation.

### **Long Paths and Memory Constraints**
- **Handled Path Length:** Ensures paths do not exceed buffer limits to prevent crashes.
- **Memory Efficiency:** Limits directory entries processed to `MAX_ENTRIES = 100`.

**Review by Nish7 — COMMENTED** [review]
Screenshot 2025-02-13 at 11 25 45 PM
I was trying to run the pushed code. is the `???` expected?

**Comment by Rocketeer8** [comment]
> Screenshot 2025-02-13 at 11 25 45 PM I was trying to run the pushed code. is the `???` expected?

Those were originally meant to represent group, owner, and date/time. But I don't think xv6 provides those without modifying the kernel. I'll clarify this in the documentation If this is ok
